packages/mui-system/src/useMediaQuery/useMediaQuery.ts

```diff
@@ -110,54 +92,7 @@ function useMediaQueryNew(
       },
     ];
   }, [getDefaultSnapshot, matchMedia, query]);
-  const match = maybeReactUseSyncExternalStore(subscribe, getSnapshot, getServerSnapshot);
-
-  return match;
-}
-
-export default function useMediaQuery<Theme = unknown>(
-  queryInput: string | ((theme: Theme) => string),
-  options: UseMediaQueryOptions = {},
-): boolean {
-  const theme = useTheme<Theme>();
-  // Wait for jsdom to support the match media feature.
-  // All the browsers MUI support have this built-in.
-  // This defensive check is here for simplicity.
-  // Most of the time, the match media logic isn't central to people tests.
-  const supportMatchMedia =
-    typeof window !== 'undefined' && typeof window.matchMedia !== 'undefined';
-  const {
-    defaultMatches = false,
-    matchMedia = supportMatchMedia ? window.matchMedia : null,
-    ssrMatchMedia = null,
-    noSsr = false,
-  } = getThemeProps({ name: 'MuiUseMediaQuery', props: options, theme });
-
-  if (process.env.NODE_ENV !== 'production') {
-    if (typeof queryInput === 'function' && theme === null) {
-      console.error(
-        [
-          'MUI: The `query` argument provided is invalid.',
-          'You are providing a function without a theme in the context.',
-          'One of the parent elements needs to use a ThemeProvider.',
-        ].join('\n'),
-      );
-    }
-  }
-
-  let query = typeof queryInput === 'function' ? queryInput(theme) : queryInput;
-  query = query.replace(/^@media( ?)/m, '');
-
-  // TODO: Drop `useMediaQueryOld` and use  `use-sync-external-store` shim in `useMediaQueryNew` once the package is stable
```

Comment:
I don't quite understand this line, is `useMediaQueryOld` bigger than `use-sync-external-store`?

Reply:
`useMediaQueryOld ` is smaller. But `use-sync-external-store` is quite small(*), and it's likely already present in our user's bundles. which in this case means us depending on it doesn't add extra cost.

(*)
<img width="420" alt="Screenshot 2024-08-28 at 05 13 58" src="https://github.com/user-attachments/assets/56beeb56-7484-440a-94a5-e2d064de6e80">

Reply:
One solution I was pondering is to provide a shim for older older React APIs. Users would have to manually import a module in the top-level if they want to use modern `@mui/material` on outdated `react`. e.g.

```tsx
// import this before anything else in user code if you're on legacy react
import `@mui/shims/setup-react-17`
```

implementation:

```tsx
// we import this where we need `useSyncExternalStore`
import useSyncExternalStore from '@mui/shims/useSyncExternalStore'
```
```tsx
// @mui/shims/setup-react-17
import { useSyncExternalStore } from 'use-sync-external-store'
globalThis._muiShims ?= {}
globalThis._muiShims.useSyncExternalStore = useSyncExternalStore
```

```tsx
// @mui/shims/useSyncExternalStore
import * as React from 'react'
import invariant from 'invariant'
const _React = { ...React }
const useSyncExternalStore = _React.useSyncExternalStore || globalThis._muiShims?.useSyncExternalStore
invariant(useSyncExternalStore, "MUI shims not installed, If you're on legacy React, please follow the guide at https://mui.com/...")
export { useSyncExternalStore }
```

That way you only pay the cost of a shim if you choose to stay on react 17. Have to say though that it's probably way overkill just to save 0.5kB.